src/Kestrel/ServerRequest.cs

```diff
@@ -231,7 +233,53 @@ bool IHttpUpgradeFeature.IsUpgradableRequest
                 }
             }
             _frame.ProduceStart();
-            return _frame.DuplexStream;
+            return Task.FromResult<Stream>(_frame.DuplexStream);
+        }
+
+        IPAddress IHttpConnectionFeature.RemoteIpAddress
+        {
+            get
+            {
+                IPAddress ip;
+                var result = IPAddress.TryParse(_frame.RemoteIpAddress, out ip);
+                return result ? ip : IPAddress.Any;
+            }
+            set
+            {
+                throw new NotSupportedException();
+            }
+        }
+
+        IPAddress IHttpConnectionFeature.LocalIpAddress
+        {
+            get
+            {
+                IPAddress ip;
+                var result = IPAddress.TryParse(_frame.LocalIpAddress, out ip);
+                return result ? ip : IPAddress.Any;
+            }
+            set
+            {
+                throw new NotSupportedException();
+            }
+        }
+
+        int IHttpConnectionFeature.RemotePort { get; set; }
+
+        int IHttpConnectionFeature.LocalPort { get; set; }
+
+        bool IHttpConnectionFeature.IsLocal
+        {
+            get
+            {
+                IPAddress ip;
+                var result = IPAddress.TryParse(_frame.RemoteIpAddress, out ip);
+                return result ? IPAddress.IsLoopback(ip) : false;
```

Comment:
You can non-loopback local connections. Elsewhere we've used `LocalIpAddress.Equals(RemoteIpAddress)`